xline/src/request_validation.rs

```diff
@@ -0,0 +1,414 @@
+        }
+        if self.ignore_value && !self.value.is_empty() {
+            return Err(ValidationError::new(
+                "ignore value is set but value is provided",
+            ));
+        }
+        if self.ignore_lease && self.lease != 0 {
+            return Err(ValidationError::new(
+                "ignore lease is set but lease is provided",
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl RequestValidator for DeleteRangeRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        if self.key.is_empty() {
+            return Err(ValidationError::new("key is not provided"));
+        }
+
+        Ok(())
+    }
+}
+
+impl RequestValidator for TxnRequest {
+    fn validation(&self) -> Result<(), ValidationError> {
+        let opc = self
+            .compare
+            .len()
+            .max(self.success.len())
+            .max(self.failure.len());
+        if opc > DEFAULT_MAX_TXN_OPS {
+            return Err(ValidationError::new("too many operations in txn request"));
+        }
+        for c in &self.compare {
+            if c.key.is_empty() {
+                return Err(ValidationError::new("key is not provided"));
+            }
+        }
+        for op in self.success.iter().chain(self.failure.iter()) {
+            if let Some(ref request) = op.request {
+                match *request {
+                    Request::RequestRange(ref r) => r.validation(),
+                    Request::RequestPut(ref r) => r.validation(),
+                    Request::RequestDeleteRange(ref r) => r.validation(),
+                    Request::RequestTxn(ref r) => r.validation(),
+                }?;
+            } else {
+                return Err(ValidationError::new("request not provided in operation"));
+            }
+        }
+
+        let _ignore_success = check_intervals(&self.success)?;
+        let _ignore_failure = check_intervals(&self.failure)?;
+
+        Ok(())
+    }
+}
+
+/// Check if puts and deletes overlap
+fn check_intervals(ops: &[RequestOp]) -> Result<(HashSet<&[u8]>, Vec<KeyRange>), ValidationError> {
+    // TODO: use interval tree is better?
+
+    let mut dels = Vec::new();
+
+    for op in ops {
+        if let Some(Request::RequestDeleteRange(ref req)) = op.request {
+            // collect dels
+            let del = KeyRange::new(req.key.as_slice(), req.range_end.as_slice());
+            dels.push(del);
+        }
+    }
+
+    let mut puts: HashSet<&[u8]> = HashSet::new();
+
+    for op in ops {
+        if let Some(Request::RequestTxn(ref req)) = op.request {
+            // handle child txn request
+            let (success_puts, mut success_dels) = check_intervals(&req.success)?;
+            let (failure_puts, mut failure_dels) = check_intervals(&req.failure)?;
+
+            for k in &success_puts {
+                if !puts.insert(k) {
+                    return Err(ValidationError::new("duplicate key given in txn request"));
```

Comment:
It's not coverred in the test. Please double check to make sure that every error case is covered in UT.

Reply:
now it should be covered by 38eab5e0dfb759ba5d36c646f05331f091231172